datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -579,26 +652,199 @@ where
+                            block_hash,
+                        ));
+                    }
+                }
+                future::ok(traces)
+            })
+            .map(move |traces| {
+                traces
+                    .iter()
+                    .filter(|trace| {
+                        let is_call = match trace.action {
+                            Action::Call(_) => true,
+                            _ => false,
+                        };
+                        if !is_call || trace.result.is_none() || trace.error.is_some() {
+                            return false;
+                        }
+                        true
+                    })
+                    .map(EthereumCall::from)
+                    .collect()
+            });
+        Box::new(calls)
+    }
+
+    fn blocks_with_triggers(
+        &self,
+        logger: &Logger,
+        from: u64,
+        to: u64,
+        log_filter_opt: Option<EthereumLogFilter>,
+        call_filter_opt: Option<EthereumCallFilter>,
+        block_filter_opt: Option<EthereumBlockFilter>,
+    ) -> Box<Future<Item = Vec<EthereumBlockPointer>, Error = Error> + Send> {
+        // If no filters are provided, return an empty vector of blocks.
+        if log_filter_opt.is_none() && call_filter_opt.is_none() && block_filter_opt.is_none() {
+            return Box::new(future::ok(vec![]));
+        }
+
+        // Each trigger filter needs to be queried for the same block range
+        // and the blocks yielded need to be deduped. If any error occurs
+        // while searching for a trigger type, the entire operation fails.
+        let eth = self.clone();
+        let mut block_futs: Vec<
+            Box<Future<Item = Vec<EthereumBlockPointer>, Error = Error> + Send>,
+        > = vec![];
+        if block_filter_opt.is_some() && block_filter_opt.clone().unwrap().trigger_every_block {
+            // All blocks in the range contain a trigger
+            block_futs.push(eth.blocks(&logger, from, to));
+        } else {
+            // Scan the block range from triggers to find relevant blocks
+            if log_filter_opt.is_some() {
+                block_futs.push(Box::new(eth.blocks_with_logs(
+                    &logger,
+                    from,
+                    to,
+                    log_filter_opt.unwrap(),
+                )));
+            }
+            if call_filter_opt.is_some() {
+                block_futs.push(Box::new(eth.blocks_with_calls(
+                    &logger,
+                    from,
+                    to,
+                    call_filter_opt.unwrap(),
+                )));
+            }
+            if block_filter_opt.is_some() {
+                let block_filter = block_filter_opt.unwrap();
+                match block_filter.contract_addresses.len() {
+                    0 => (),
+                    _ => {
+                        let call_filter = EthereumCallFilter::from(block_filter);
```

Comment:
Can we add a comment here about why we are using a call filter for the block here?